calodiffusion/models/sample.py

```diff
@@ -1121,3 +1121,152 @@ def __call__(self, model, start, energy, layers, num_steps, sample_offset, debug
         
         return self.sampler(start, debug=debug, offset=sample_offset)
 
+class PushforwardTraining(Sample): 
+    """Pushforward sampling to be paired with IMM loss - https://arxiv.org/abs/2503.07565"""
+    def __init__(self, config):
+        super().__init__(config)
+        self.epsilon = self.sample_config.get("EPSILON", 0.01)
+        self.noise_low = self.sample_config.get("NOISE_LOW", 0.0)
+        self.noise_high = self.sample_config.get("NOISE_HIGH", 0.994)
+
+        self.t_distribution = torch.distributions.LogNormal(
+            self.sample_config.get("P_MEAN", 0.0),
+            self.sample_config.get("P_STD", 0.1)
+        )
+
+        self.s_distribution = torch.distributions.Uniform(
+            self.noise_low, 
+            self.noise_high
+        )
+
+        self.k_noise_seperation = self.sample_config.get("K", 12)
+        self.noise_schedule = self.sample_config.get("NOISE_SCHEDULE", "fm")
+
+    def compute_noise_offset(self, noise_sampled): 
+        "Convert s -> r, ensuring there is a proper offset between the two sampling parameters"
+
+        u = (self.noise_high - self.noise_low) * (1/2) ** self.k_noise_seperation
+        return (noise_sampled - u).clamp(min=self.noise_low, max=self.noise_high) 
+
+    def noise_ratio_to_time(self, sample): 
+        """Convert the sample of the noise to a timestep """
+
+        if self.noise_schedule == "vp_cosine":
+            t = torch.arctan(sample) / (torch.pi * 0.5) 
+
+        elif self.noise_schedule == "fm":
+            t = sample / (1 + sample)
+        
+        # p(nan) := 1 
+        t = torch.nan_to_num(t, nan=1)
+        return t
+
+    def get_alpha_sigma(self, time_step): 
+        if self.noise_schedule == "fm":
+            # Flow matching: linear interpolation between clean and noise
+            alpha = 1 - time_step  # Signal coefficient decreases linearly
+            sigma = time_step      # Noise coefficient increases linearly
+            
+        elif self.noise_schedule == "vp_cosine":
+            # VP cosine schedule: smooth transition using cosine/sine
+            alpha = torch.cos(time_step * torch.pi * 0.5)
+            sigma = torch.sin(time_step * torch.pi * 0.5)
+
+        return alpha, sigma
+
+    def sample_noise(self, x) -> tuple[torch.Tensor, torch.Tensor]: 
+        """
+        Time sampling that produces T and R samples (high noise and low noise).
```

Comment:
it would be a useful reference to have a central comment defining/explaining t, s, r in the IMM context